File: vol2/ch21/LeetCode174.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        INF=0x3f3f3f3f
        m,n=len(dungeon),len(dungeon[0])
        dp=[[INF for _ in range(0,n+1)] for _ in range(0,m+1)]
        dp[m][n-1],dp[m-1][n]=1,1
        for i in range(m-1,-1,-1):
            for j in range(n-1,-1,-1):
                mind=min(dp[i+1][j],dp[i][j+1])
                dp[i][j]=max(mind-dungeon[i][j],1)
        return dp[0][0]

File: vol2/ch21/LeetCode174.cpp
class Solution {
    const int INF=0x3f3f3f3f;
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int m=dungeon.size(),n=dungeon[0].size();
        vector<vector<int>> dp(m+1,vector<int>(n+1,INF));
        dp[m][n-1]=dp[m-1][n]=1;
        for (int i=m-1;i>=0; i--) {
            for (int j=n-1;j>=0;j--) {
                int mind=min(dp[i+1][j],dp[i][j+1]);
                dp[i][j]=max(mind-dungeon[i][j],1);
            }
        }
        return dp[0][0];
    }
};

